### PR TITLE
feat(DICOM Upload): Added new DICOM upload dialogue launched from the worklist page

### PR DESCRIPTION
**Pull request by jbocce**
### Context

See OHIF #3297

### Changes & Results

- DicomWebDataSource.store.dicom now accepts an ArrayBuffer of data sets to store
- DicomWebDataSource.store.dicom now also accepts optional callbacks to track upload and an AbortSignal object to cancel upload
- Added DicomFileUploader class that performs and tracks the upload of a DICOM file
- Added various UI pieces for the upload: DicomUpload, DicomUploadProgress, DicomUploadProgressItem
- ProgressLoadingBar was extracted from LoadingIndicatorProgress so it can be reused
- Modal dialogues can now optionally prevent an outside click from closing the Modal

### Testing

N.B. Accurate progress of upload and the ability to cancel an upload requires a more recent version of `dicomweb-client`. Without a more recent version, progress is limited to each file completing entirely and cancel is NOT possible.

Also note that upload requires a data source that allows it and has the configuration property `dicomUploadEnabled` set to true. At present, both the `local_orthanc.js` and `local_dcm4chee.js` configuration have upload enabled, so testing should proceed with those.

1. From the worklist page, click the Upload link/button.
2. A modal dialogue should appear prompting to upload DICOM files either via: drag and drop, selecting files individually or selecting an entire folder. Note that the modal dialogue can be closed with a click of the X button in the top right corner.
3. Once DICOM files are selected/dropped, the upload process begins. 
    - The X to close the dialogue should disappear
    - A summary of the number of files completed (i.e. either errored, cancelled or successful) should be displayed. 
    - To the right of that number, an estimate to the time remaining will display whenever/if one is calculated.
    - A link/button to cancel all uploads
    - A progress bar that shows the infinite (i.e. back and forth) progress until at least one percent is processed or one pixel of progress can be depicted (whichever comes first)
    - The numerical percentage appears to the right of the progress bar
    - An error indicator icon will appear to the right of the numerical percentage. Clicking this icon will toggle between filtering and only showing errors and all file upload results/progress
    - A scrollable list, listing each file and its progress appears below the progress bar section/row. From the left, the status, file name, percentage complete and an X icon to cancel an individual file upload should appear in each row. The cancel button is only available prior and during the file's upload but is remove after the file is completed
4. Once all files are complete, A completed message appears and a button to return to the viewer/worklist will display. Clicking this button will return to the worklist and the worklist should automatically refresh to list the uploaded studies.

Things to consider while testing:
- try each method of choosing files (i.e. drag and drop, individual files and folders)
- vary the upload speed using the browser developer tools
- cancel various requests and all requests

#### PR

- [x] My Pull Request title is descriptive, accurate and follows the
  semantic-release format and guidelines.

#### Code

- [x] My code has been well-documented (function documentation, inline comments,
  etc.)

#### Public Documentation Updates

- [x] The documentation page has been updated as necessary for any public API
  additions or removals.

#### Tested Environment

- [x] OS: Windows 11
- [x] Node version: 16.14.0 
- [x] Chrome: 112.0.5615.49
  

[blog]: https://circleci.com/blog/triggering-trusted-ci-jobs-on-untrusted-forks/
[script]: https://github.com/jklukas/git-push-fork-to-upstream-branch

### CODE REVIEW

File: extensions/default/src/DicomWebDataSource/index.js [hunk 1]
```diff
@@ -219,36 +219,47 @@ function createDicomWebApi(dicomWebConfig, userAuthenticationService) {
     },
 
     store: {
-      dicom: async dataset => {
+      dicom: async (dataset, uploadCallbacks = {}, abortSignal) => {
```
Comment (sedghi) [inline extensions/default/src/DicomWebDataSource/index.js:222]: `abortSignal` should have default if it is coming after the second one that has a default

File: platform/ui/src/components/StudyListFilter/StudyListFilter.tsx [hunk 2]
```diff
@@ -11,6 +11,7 @@ const StudyListFilter = ({
   clearFilters,
   isFiltering,
   numOfStudies,
+  onUploadClick,
```
Comment (sedghi) [inline platform/ui/src/components/StudyListFilter/StudyListFilter.tsx:14]: `onUploadClick` sounds to me like an eventHandler, an action, but seems like it is a flag to render ICON?

Comment (sedghi) [reply platform/ui/src/components/StudyListFilter/StudyListFilter.tsx:14]: nevermind

File: platform/viewer/src/routes/WorkList/WorkList.tsx [hunk 3]
```diff
@@ -429,6 +432,34 @@ function WorkList({
...
+    title: 'Upload files',
+    closeButton: true,
+    shouldCloseOnEsc: false,
+    shouldCloseOnOverlayClick: false,
+    content: DicomUpload.bind(null, {
+      dataSource,
+      onComplete: handleUploadComplete,
+      onStarted: handleUploadStarted,
+    }),
+  };
```
Comment (sedghi) [inline platform/viewer/src/routes/WorkList/WorkList.tsx:457]: I recommend reducing the lines of code with somethign like this, which is more readable to me too

```js
const uploadProps = {
  title: "Upload files",
  closeButton: true,
  shouldCloseOnEsc: false,
  shouldCloseOnOverlayClick: false,
  content: DicomUpload.bind(null, {
    dataSource,
    onComplete: () => {
      hide()
      onRefresh()
    },
    onStarted: () => {
      show({
        ...uploadProps,
        closeButton: false,
      })
    },
  }),
}
```

File: platform/viewer/src/routes/WorkList/WorkList.tsx [hunk 4]
```diff
@@ -445,6 +476,7 @@ function WorkList({
           onChange={setFilterValues}
           clearFilters={() => setFilterValues(defaultFilterValues)}
           isFiltering={isFiltering(filterValues, defaultFilterValues)}
+          onUploadClick={handleShowUpload}
```
Comment (sedghi) [inline platform/viewer/src/routes/WorkList/WorkList.tsx:479]: also here you can just say

```js
onUploadClick = {() => show(uploadProps)}
```

File: platform/viewer/src/routes/DataSourceWrapper.tsx [hunk 5]
```diff
@@ -125,6 +126,7 @@ function DataSourceWrapper(props) {
       dataTotal={data.total}
       dataSource={dataSource}
       isLoadingData={isLoading}
+      onRefresh={() => setData(DEFAULT_DATA)}
```
Comment (sedghi) [inline platform/viewer/src/routes/DataSourceWrapper.tsx:129]: can you please add a comment here of why we are doing this setData on refresh?

File: platform/viewer/src/components/DicomUpload/DicomUploadProgress.tsx [hunk 6]
```diff
@@ -0,0 +1,407 @@
+import React, { useCallback, useEffect, useRef, useState } from 'react';
+import { ReactElement } from 'react';
```
Comment (sedghi) [inline platform/viewer/src/components/DicomUpload/DicomUploadProgress.tsx:2]: merge with above

File: platform/viewer/src/components/DicomUpload/DicomUploadProgress.tsx [hunk 7]
```diff
@@ -0,0 +1,407 @@
...
+import PropTypes from 'prop-types';
+import { Button, Icon, ProgressLoadingBar } from '@ohif/ui';
+import DicomFileUploader, {
+  EVENTS,
+  UploadStatus,
+} from '../../utils/DicomFileUploader';
+import { DicomFileUploaderProgressEvent } from '../../utils/DicomFileUploader';
```
Comment (sedghi) [inline platform/viewer/src/components/DicomUpload/DicomUploadProgress.tsx:9]: merge with above

File: platform/viewer/src/components/DicomUpload/DicomUploadProgress.tsx [hunk 8]
```diff
@@ -0,0 +1,407 @@
...
+import DicomUploadProgressItem from './DicomUploadProgressItem';
+import classNames from 'classnames';
+
+export type DicomUploadProgressProps = {
```
Comment (sedghi) [inline platform/viewer/src/components/DicomUpload/DicomUploadProgress.tsx:13]: it is getting exported but not used anywhere except this file

File: platform/viewer/src/components/DicomUpload/DicomUploadProgress.tsx [hunk 9]
```diff
@@ -0,0 +1,407 @@
...
+// of the very first interval to get a reasonable estimate on screen in
+// a reasonable amount of time. The length of each interval after the first
+// is based on the upload rate calculated. Faster rates use this base interval
+// length. Slower rates below UPLOAD_RATE_THRESHOLD get longer interval times
+// to obtain more accurate upload rates.
+const BASE_INTERVAL_TIME = 15000;
+
+// The upload rate threshold to determine the length of the interval to
+// calculate the upload rate.
+const UPLOAD_RATE_THRESHOLD = 75;
```
Comment (sedghi) [inline platform/viewer/src/components/DicomUpload/DicomUploadProgress.tsx:30]: I guess the whole dicomUpload should move to the cornerstone extension

1. it is using cornersone-wado-image-loader
2. There are configurations that can be done apparently and it makes sense to be inisde extension than core

I'm not sure how do we handle the worklist page though which is in the core. That should go to default extension in my opinion. Same for Local it should be in the default (but not the concern of this PR). It is odd that the viewer layout (grid, panel, toolbar) are all in the extensions and we have worklist in the core

File: platform/viewer/src/components/DicomUpload/DicomUploadProgress.tsx [hunk 10]
```diff
@@ -0,0 +1,407 @@
...
+        currentUploadSizeRef.current - intervalStartUploadSize;
+
+      const now = Date.now();
+      const timeSinceStartOfInterval = now - intervalStartTime;
+
+      // Calculate and set the upload rate (ref)
+      uploadRateRef.current =
+        uploadSizeFromStartOfInterval / timeSinceStartOfInterval;
+
+      console.info(uploadRateRef.current);
```
Comment (sedghi) [inline platform/viewer/src/components/DicomUpload/DicomUploadProgress.tsx:86]: do we need this?

File: platform/viewer/src/components/DicomUpload/DicomUploadProgress.tsx [hunk 11]
```diff
@@ -0,0 +1,407 @@
...
+    } else if (timeRemaining < 3600000) {
+      const minutesRemaining = Math.ceil(timeRemaining / 60000);
+      return `${minutesRemaining} ${
+        minutesRemaining === 1 ? 'minute' : 'minutes'
+      }`;
+    } else {
+      const hoursRemaining = Math.ceil(timeRemaining / 3600000);
+      return `${hoursRemaining} ${hoursRemaining === 1 ? 'hour' : 'hours'}`;
+    }
+  };
```
Comment (sedghi) [inline platform/viewer/src/components/DicomUpload/DicomUploadProgress.tsx:128]: I recommend to move this before or after useEffects, not in the middle of them

File: platform/viewer/src/components/DicomUpload/DicomUploadProgress.tsx (hunk 9 above)
Comment (jbocce) [reply platform/viewer/src/components/DicomUpload/DicomUploadProgress.tsx:39]: I agree and will look into moving the DICOM upload pieces into the cornerstone extension and referencing it via a customization module.

**Comment by jbocce** [comment]
@sedghi, thank you for your review. Your suggestions cleaned up my code nicely. I still have to move the DICOM upload bits out of `viewer` and I am still looking into enabling/disabling the upload via configuration per data source.

File: platform/viewer/src/components/DicomUpload/DicomUploadProgress.tsx (hunk 9 above)
Comment (jbocce) [reply platform/viewer/src/components/DicomUpload/DicomUploadProgress.tsx]: Ok the most recent push includes a commit that takes care of this.

**Comment by jbocce** [comment]
@wayfarer3130 , the DICOM upload UI is added via a customization module from the cornerstone extension. I am including you  on the PR specifically for that piece, but feel free to comment on other code as well. Thank you kindly.

File: platform/viewer/src/routes/WorkList/WorkList.tsx [hunk 12]
```diff
@@ -29,6 +29,8 @@ import {
 } from '@ohif/ui';
 
 import i18n from '@ohif/i18n';
+import DicomUpload from '@components/DicomUpload/DicomUpload';
```
Comment (sedghi) [inline platform/viewer/src/routes/WorkList/WorkList.tsx:32]: I don't think you need this anymore right?

File: extensions/cornerstone/src/components/DicomUpload/DicomUpload.tsx [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+  const onDrop = useCallback(async acceptedFiles => {
+    onStarted();
+    setDicomFileUploaderArr(
+      acceptedFiles.map(file => new DicomFileUploader(file, dataSource))
+    );
+  }, []);
+
+  return (
+    <div className="h-[calc(100vh-300px)] min-h-[250px] flex flex-col bg-black">
+      {dicomFileUploaderArr.length ? (
```
Comment (sedghi) [inline extensions/cornerstone/src/components/DicomUpload/DicomUpload.tsx:31]: I would recommend refactoring this into couple of components, or functions to make it easier to read something like 

```js
<div className="h-[calc(100vh-300px)] min-h-[250px] flex flex-col bg-black">
  {dicomFileUploaders.length ? (
    <DicomUploadProgress
      dicomFileUploaderArr={Array.from(dicomFileUploaders)}
      onComplete={onComplete}
    />
  ) : (
    renderDropzone()
  )}
</div>
```

**Review by sedghi — COMMENTED** [review]
Looks great, minor comments

**Comment by sedghi** [comment]
There is an unexpected layout shift for the file name when the file names are large. During upload it is .... then after the file extension reveals and cause a layout shift

image

---

User should not be able to highlight the cancel button

image

**Review by sedghi — APPROVED** [review]
Great work indeed, thanks

**Pull request merged**